radicle-tui/src/ui/widget.rs

```diff
@@ -0,0 +1,90 @@
+use std::ops::Deref;
+
+use tuirealm::command::{Cmd, CmdResult};
+use tuirealm::props::{AttrValue, Attribute, Color, Props};
+use tuirealm::tui::layout::Rect;
+use tuirealm::{Frame, MockComponent, State};
+
+pub type BoxedWidget<T> = Box<Widget<T>>;
+
+pub trait WidgetComponent {
+    fn view(&mut self, properties: &Props, frame: &mut Frame, area: Rect);
+
+    fn state(&self) -> State;
+
+    fn perform(&mut self, cmd: Cmd) -> CmdResult;
+}
+
+#[derive(Clone)]
+pub struct Widget<T: WidgetComponent> {
+    component: T,
+    properties: Props,
+}
+
+impl<T: WidgetComponent> Deref for Widget<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.component
+    }
+}
+
+impl<T: WidgetComponent> Widget<T> {
+    pub fn new(component: T) -> Self {
+        Widget {
+            component,
+            properties: Props::default(),
+        }
+    }
+
+    pub fn foreground(mut self, fg: Color) -> Box<Self> {
+        self.attr(Attribute::Foreground, AttrValue::Color(fg));
+        self.to_boxed()
+    }
+
+    pub fn highlight(mut self, fg: Color) -> Box<Self> {
+        self.attr(Attribute::HighlightedColor, AttrValue::Color(fg));
+        self.to_boxed()
+    }
+
+    pub fn background(mut self, bg: Color) -> Box<Self> {
+        self.attr(Attribute::Background, AttrValue::Color(bg));
+        self.to_boxed()
```

Comment:
In all these functions we should just return `Self`, then boxing is explicit when you call `to_boxed`